--- /home/ncl/Downloads/android_kernel_xiaomi_rosemary/drivers/misc/mediatek/base/power/mt6785/mtk_gpufreq_core.c	2025-06-28 13:18:53.295872047 +0700
+++ /home/ncl/Downloads/realmeC11_realmeC12_realmeC15_AndroidQ-kernel-source/drivers/misc/mediatek/base/power/mt6785/mtk_gpufreq_core.c	2025-08-06 09:24:52.653743142 +0700
@@ -107,6 +107,9 @@
 static void __mt_gpufreq_low_batt_protect(unsigned int limited_index);
 static void __mt_update_gpufreqs_power_table(void);
 static void __mt_gpufreq_update_max_limited_idx(void);
+#ifdef VENDOR_EDIT
+static void __mt_gpufreq_update_min_limited_idx(void);
+#endif
 static unsigned int __mt_gpufreq_calculate_dds(unsigned int freq_khz,
 		enum g_post_divider_power_enum post_divider_power);
 static void __mt_gpufreq_setup_opp_power_table(int num);
@@ -189,6 +192,9 @@
 static unsigned int g_fixed_freq;
 static unsigned int g_fixed_volt;
 static unsigned int g_max_limited_idx;
+#ifdef VENDOR_EDIT
+static unsigned int g_min_limited_idx = INT_MAX;
+#endif
 static unsigned int g_pbm_limited_power;
 static unsigned int g_thermal_protect_power;
 static unsigned int g_vgpu_sfchg_rrate;
@@ -222,6 +228,9 @@
 static DEFINE_MUTEX(mt_gpufreq_lock);
 static DEFINE_MUTEX(mt_gpufreq_power_lock);
 static unsigned int g_limited_idx_array[NUMBER_OF_LIMITED_IDX] = { 0 };
+#ifdef VENDOR_EDIT
+static unsigned int g_limited_min_idx_array[NR_IDX_POWER_MIN_LIMITED] = { INT_MAX };
+#endif
 static bool g_limited_ignore_array[NUMBER_OF_LIMITED_IDX] = { false };
 static void __iomem *g_apmixed_base;
 static void __iomem *g_efuse_base;
@@ -233,16 +242,93 @@
  * ===============================================
  */
 
+#ifdef VENDOR_EDIT
+int mt_gpufreq_scene_protect(unsigned int min_freq, unsigned int max_freq)
+{
+	int i = 0;
+	bool min_done = false;
+	bool max_done = false;
+
+	if (min_freq > max_freq) {
+		gpufreq_perr("@%s: GPU DVFS invalid input min_freq:%u max_freq:%u\n",
+			__func__, min_freq, max_freq);
+		return -EINVAL;
+	}
+
+	mutex_lock(&mt_gpufreq_lock);
+	for (i = 0; i < g_opp_idx_num; i++) {
+		//max freq
+		if (g_opp_table[i].gpufreq_khz <= max_freq && !max_done) {
+			g_limited_idx_array[IDX_SCENE_LIMITED] = i;
+			max_done = true;
+		}
+		//min freq
+		if (g_opp_table[g_opp_idx_num - 1 - i].gpufreq_khz >= min_freq && !min_done) {
+			g_limited_min_idx_array[IDX_SCENE_MIN_LIMITED] = g_opp_idx_num - 1 - i;
+			min_done = true;
+		}
+	}
+
+	if (!max_done)
+		g_limited_idx_array[IDX_SCENE_LIMITED] = 0;
+	if (!min_done)
+		g_limited_min_idx_array[IDX_SCENE_MIN_LIMITED] = g_opp_idx_num - 1;
+
+	// if min freq > max freq
+	if (g_limited_min_idx_array[IDX_SCENE_MIN_LIMITED] <
+		g_limited_idx_array[IDX_SCENE_LIMITED]) {
+		g_limited_idx_array[IDX_SCENE_LIMITED] = g_limited_min_idx_array[IDX_SCENE_MIN_LIMITED];
+	}
+
+	__mt_gpufreq_update_max_limited_idx();
+	__mt_gpufreq_update_min_limited_idx();
+	mutex_unlock(&mt_gpufreq_lock);
+	return 0;
+}
+EXPORT_SYMBOL(mt_gpufreq_scene_protect);
+#endif /* VENDOR_EDIT */
 /*
  * API : handle frequency change request
  */
+#ifdef VENDOR_EDIT
+static u64 mt_gpufreq_time_in_state[32] = {0};
+unsigned int time_in_state_run = 0;
+u64 prev_switch_time = 0;
+
+static int mt_gpufreq_in_time_proc_show(struct seq_file *m, void *v)
+{
+	int i;
+
+	//gpufreq from high to low
+	for (i = 0; i < g_opp_idx_num; i++) {
+		seq_printf(m, "%llu ", mt_gpufreq_time_in_state[i] / 1000);
+	}
+	seq_printf(m, "\n");
+
+	return 0;
+}
+
+static int mt_gpufreq_opp_list_proc_show(struct seq_file *m, void *v)
+{
+	int i;
+
+	for (i = 0; i < g_opp_idx_num; i++) {
+		seq_printf(m, "%d ", g_opp_table[i].gpufreq_khz);
+	}
+	seq_printf(m, "\n");
+
+	return 0;
+}
+#endif /* VENDOR_EDIT */
 unsigned int mt_gpufreq_target(unsigned int idx)
 {
 	unsigned int target_freq;
 	unsigned int target_volt;
 	unsigned int target_idx;
 	unsigned int target_cond_idx;
-
+#ifdef VENDOR_EDIT
+	u64 now;
+#endif
 	mutex_lock(&mt_gpufreq_lock);
 
 #ifdef MT_GPUFREQ_OPP_STRESS_TEST
@@ -284,6 +370,22 @@
 		}
 	}
 
+#ifdef VENDOR_EDIT
+	//min freq
+	gpufreq_pr_debug("@%s: OPP freq is limited by Thermal/Power/PBM, g_min_limited_idx = %d g_opp_idx_num = %d\n",
+					__func__, target_cond_idx, g_opp_idx_num);
+	if (g_min_limited_idx != g_opp_idx_num - 1) {
+		if (target_freq < g_opp_table[g_min_limited_idx].gpufreq_khz) {
+			target_freq = g_opp_table[g_min_limited_idx].gpufreq_khz;
+			target_volt = g_opp_table[g_min_limited_idx].gpufreq_volt;
+			target_idx = g_opp_table[g_min_limited_idx].gpufreq_idx;
+			target_cond_idx = g_min_limited_idx;
+			gpufreq_pr_debug("@%s: OPP freq is limited by Thermal/Power/PBM, g_min_limited_idx = %d\n",
+					__func__, target_cond_idx);
+		}
+	}
+#endif /* VENDOR_EDIT */
+
 	/* If /proc command keep OPP freq */
 	if (g_keep_opp_freq_state) {
 		target_freq = g_opp_table[g_keep_opp_freq_idx].gpufreq_khz;
@@ -331,6 +433,18 @@
 	__mt_gpufreq_set(g_cur_opp_freq, target_freq, g_cur_opp_volt,
 		target_volt, g_cur_opp_vsram_volt,
 		__mt_gpufreq_get_vsram_volt_by_target_volt(target_volt));
+#ifdef VENDOR_EDIT
+	now = local_clock();
+	if (likely(time_in_state_run == 1)) {
+		if (now > prev_switch_time && target_idx < min((unsigned int)32, g_opp_idx_num)) {
+			mt_gpufreq_time_in_state[target_idx] += now - prev_switch_time;
+			prev_switch_time = now;
+		}
+	} else {
+		time_in_state_run = 1;
+		prev_switch_time = now;
+	}
+#endif /* VENDOR_EDIT */
 
 	g_cur_opp_idx = target_idx;
 	g_cur_opp_cond_idx = target_cond_idx;
@@ -558,6 +672,15 @@
 {
 	return g_opp_idx_num;
 }
+#ifdef VENDOR_EDIT
+EXPORT_SYMBOL(mt_gpufreq_get_dvfs_table_num);
+/* API : get OPP table */
+struct g_opp_table_info *mt_gpufreq_get_dvfs_table(void)
+{
+	return g_opp_table;
+}
+EXPORT_SYMBOL(mt_gpufreq_get_dvfs_table);
+#endif /* VENDOR_EDIT */
 
 /* API : get frequency via OPP table index */
 unsigned int mt_gpufreq_get_freq_by_idx(unsigned int idx)
@@ -635,6 +758,21 @@
 	return g_max_limited_idx;
 }
 
+#ifdef VENDOR_EDIT
+unsigned int mt_gpufreq_get_thermal_limit_max_index(void)
+{
+	return mt_gpufreq_get_thermal_limit_index();
+}
+EXPORT_SYMBOL(mt_gpufreq_get_thermal_limit_max_index);
+
+unsigned int mt_gpufreq_get_thermal_limit_min_index(void)
+{
+	gpufreq_pr_debug("@%s: current GPU Thermal/Power/PBM limit min index is %d\n",
+			__func__, g_min_limited_idx);
+	return g_min_limited_idx;
+}
+EXPORT_SYMBOL(mt_gpufreq_get_thermal_limit_min_index);
+#endif /* VENDOR_EDIT */
 /*
  * API : get current Thermal/Power/PBM limited OPP table frequency
  */
@@ -647,6 +785,15 @@
 }
 EXPORT_SYMBOL(mt_gpufreq_get_thermal_limit_freq);
 
+#ifdef VENDOR_EDIT
+unsigned int mt_gpufreq_get_min_limit_freq(void)
+{
+	gpufreq_pr_debug("@%s: current GPU thermal limit freq is %d MHz\n",
+			__func__, g_opp_table[g_min_limited_idx].gpufreq_khz / 1000);
+	return g_opp_table[g_min_limited_idx].gpufreq_khz;
+}
+EXPORT_SYMBOL(mt_gpufreq_get_min_limit_freq);
+#endif /*VENDOR_EDIT */
 /*
  * API : get current OPP table conditional index
  */
@@ -939,6 +1086,29 @@
 
 	return 0;
 }
+#ifdef VENDOR_EDIT
+static int mt_max_freq_proc_show(struct seq_file *m, void *v)
+{
+	int max_gpu_freq;
+	int i;
+
+	i = g_max_limited_idx;
+	max_gpu_freq = g_opp_table[i].gpufreq_khz;
+	seq_printf(m, "%d\n", max_gpu_freq);
+	return 0;
+}
+
+static int mt_min_freq_proc_show(struct seq_file *m, void *v)
+{
+	int min_gpu_freq;
+	int i;
+
+	i = g_min_limited_idx;
+	min_gpu_freq = g_opp_table[i].gpufreq_khz;
+	seq_printf(m, "%d\n", min_gpu_freq);
+	return 0;
+}
+#endif /* VENDOR_EDIT */
 
 /*
  * PROCFS : show OPP power table
@@ -986,6 +1156,9 @@
 	seq_printf(m, "g_DVFS_off_by_ptpod_idx = %d\n",
 		g_DVFS_off_by_ptpod_idx);
 	seq_printf(m, "g_max_limited_idx = %d\n", g_max_limited_idx);
+#ifdef VENDOR_EDIT
+	seq_printf(m, "g_min_limited_idx = %d\n", g_min_limited_idx);
+#endif
 	seq_printf(m, "g_opp_springboard_idx = %d\n", g_opp_springboard_idx);
 	seq_printf(m, "gpu_loading = %d\n", gpu_loading);
 
@@ -995,6 +1168,13 @@
 
 	return 0;
 }
+#ifdef VENDOR_EDIT
+static int mt_cur_freq_proc_show(struct seq_file *m, void *v)
+{
+	seq_printf(m, "%d\n", g_cur_opp_freq);
+	return 0;
+}
+#endif /* VENDOR_EDIT */
 
 #ifdef MT_GPUFREQ_OPP_STRESS_TEST
 /*
@@ -1305,6 +1485,15 @@
 PROC_FOPS_RO(gpufreq_power_dump);
 PROC_FOPS_RW(gpufreq_opp_freq);
 PROC_FOPS_RO(gpufreq_var_dump);
+#ifdef VENDOR_EDIT
+PROC_FOPS_RO(gpufreq_in_time);
+PROC_FOPS_RO(gpufreq_opp_list);
+#endif
+#ifdef VENDOR_EDIT
+PROC_FOPS_RO(max_freq);
+PROC_FOPS_RO(min_freq);
+PROC_FOPS_RO(cur_freq);
+#endif /* VENDOR_EDIT */
 PROC_FOPS_RW(gpufreq_fixed_freq_volt);
 static int __mt_gpufreq_create_procfs(void)
 {
@@ -1323,6 +1512,15 @@
 		PROC_ENTRY(gpufreq_power_dump),
 		PROC_ENTRY(gpufreq_opp_freq),
 		PROC_ENTRY(gpufreq_var_dump),
+#ifdef VENDOR_EDIT
+		PROC_ENTRY(gpufreq_in_time),
+		PROC_ENTRY(gpufreq_opp_list),
+#endif /* VENDOR_EDIT */
+#ifdef VENDOR_EDIT
+		PROC_ENTRY(max_freq),
+		PROC_ENTRY(min_freq),
+		PROC_ENTRY(cur_freq),
+#endif /* VENDOR_EDIT */
 		PROC_ENTRY(gpufreq_fixed_freq_volt),
 	};
 
@@ -1926,7 +2124,7 @@
  * is 900MHz, so post
  * divider could be 2(X), 4(3600/4), 8(X), 16(X).
  * - It may have special requiremt by DE in different efuse value
- * - e.g: In mt6757, efuse value(3'b001), VCO range is 1.5GHz - 3.8GHz,
+ * - e.g: In Olympus, efuse value(3'b001), VCO range is 1.5GHz - 3.8GHz,
  * required frequency
  * range is 375MHz - 900MHz, It can only use post divider 4, no post divider 2.
  */
@@ -2141,6 +2339,23 @@
 	gpufreq_pr_debug("@%s: g_max_limited_idx = %d\n",
 		__func__, g_max_limited_idx);
 }
+#ifdef VENDOR_EDIT
+static void __mt_gpufreq_update_min_limited_idx(void)
+{
+	int i = 0;
+	unsigned limited_idx = g_opp_idx_num - 1;
+
+	for (i = 0; i < NR_IDX_POWER_MIN_LIMITED; i++) {
+		if (g_limited_min_idx_array[i] < limited_idx)
+			limited_idx = g_limited_min_idx_array[i];
+		gpufreq_pr_debug("g_limited_min_idx_array[%d] = %d\n",
+			i, g_limited_min_idx_array[i]);
+	}
+	g_min_limited_idx = limited_idx;
+	gpufreq_pr_debug("Final limit frequency lower bound to id = %d, freq = %d\n",
+			g_min_limited_idx, g_opp_table[g_min_limited_idx].gpufreq_khz);
+}
+#endif
 
 #ifdef MT_GPUFREQ_BATT_OC_PROTECT
 /*
@@ -2292,6 +2507,9 @@
 	g_opp_idx_num = num;
 	g_max_limited_idx = 0;
 
+#ifdef VENDOR_EDIT
+	g_min_limited_idx = g_opp_idx_num - 1;
+#endif /* VENDOR_EDIT */
 	__mt_gpufreq_calculate_springboard_opp_index();
 	__mt_gpufreq_setup_opp_power_table(num);
 }
